integrations/kube-agent-updater/pkg/img/cosign.go

```diff
@@ -0,0 +1,129 @@
+// SHA-256 is the de-facto standard, and we should be able to do everything with
+// it: https://github.com/opencontainers/image-spec/blob/main/descriptor.md#sha-256
+const hashAlgo = crypto.SHA256
+
+type cosignKeyValidator struct {
+	verifier        signature.Verifier
+	skid            []byte
+	name            string
+	registryOptions []ociremote.Option
+}
+
+// Name returns the validator name, it is composed of a pretty name chosen at creation
+// and its public SubjectKeyID hex-encoded.
+func (v *cosignKeyValidator) Name() string {
+	prettySKID := hex.EncodeToString(v.skid)
+	return v.name + "-" + prettySKID
+}
+
+// TODO: cache this to protect against registry quotas
+// The image validation is only invoked when we are in a maintenance window and
+// the target version is different than our current version. In regular usage we
+// are called only once per update. However, Kubernetes controllers failure mode
+// is usually infinite retry loop. If something fails after the image validation,
+// we might get called in a loop indefinitely. To mitigate the impact of such
+// failure, ValidateAndResolveDigest should cache its result.
+
+// ValidateAndResolveDigest resolves the image digest and validates it was
+// signed with cosign using a trusted static key.
+func (v *cosignKeyValidator) ValidateAndResolveDigest(ctx context.Context, image reference.NamedTagged) (NamedTaggedDigested, error) {
+	// No RegistryClientOpts for now, this means no private repo support
+	// This might also be useful for local testing (running a test registry)
+	checkOpts := &cosign.CheckOpts{
+		RegistryClientOpts: v.registryOptions,
+		Annotations:        nil,
+		ClaimVerifier:      cosign.SimpleClaimVerifier,
+		SigVerifier:        v.verifier,
+		IgnoreTlog:         true, // TODO: should we keep this?
+	}
+
+	ref, err := NamedTaggedToDigest(image)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	verified, _, err := cosign.VerifyImageSignatures(ctx, ref, checkOpts)
+	if err != nil {
+		return nil, trace.Trust(err, "failed to verify image signature")
+	}
+	if len(verified) == 0 {
+		return nil, trace.Wrap(&trace.TrustError{Message: "cannot validate image: no valid signature found"})
+	}
+
+	// There are legitimate use-cases where the signing reference is not the same
+	// as the img we're pulling from: img promoted to an internal registry,
+	// cache, registry migration, ...
+	// Thus, we take only take the digest from the signature and use it with our base img.
+	// This is what cosign.SimpleClaimVerifier is doing anyway.
+	digestedImage := NewImageRef(ref.RegistryStr(), ref.RepositoryStr(), image.Tag(), digest.Digest(ref.DigestStr()))
+	return digestedImage, nil
+}
+
+// NewCosignSingleKeyValidator takes a PEM-encoded public key and returns an
+// img.Validator that checks the image was signed with cosign by the
+// corresponding private key.
+func NewCosignSingleKeyValidator(pem, name string) (Validator, error) {
```

Comment:
pem as strings makes me always think it's the file location instead of the pem content itself.